src/seals/base_envs.py

```diff
@@ -247,6 +346,88 @@ def feature_matrix(self):
         """Matrix mapping states to feature vectors."""
         # Construct lazily to save memory in algorithms that don't need features.
         if self._feature_matrix is None:
+            # TODO(juan) Space() does not have an `n` attribute (?).
+            #  Are we hinting the wrong type?
             n_states = self.state_space.n
             self._feature_matrix = np.eye(n_states)
         return self._feature_matrix
+
+    @property
+    def state_dim(self):
+        """Number of states in this MDP (int)."""
+        return self.transition_matrix.shape[0]
+
+    @property
+    def action_dim(self) -> int:
+        """Number of action vectors (int)."""
+        return self.transition_matrix.shape[1]
+
+    @property
+    def obs_dim(self) -> int:
+        """Size of observation vectors for this MDP."""
+        return self.observation_matrix.shape[1]
+
+    @property
+    def obs_dtype(self) -> int:
+        """Data type of observation vectors (e.g. np.float32)."""
+        return self.observation_matrix.dtype
+
+
+class TabularModelPOMDP(BaseTabularModelPOMDP[np.ndarray]):
+    """Tabular model POMDP.
+
+    This class is specifically for environments where observation != state,
+    from both a typing perspective but also by defining the method that
+    draws observations from the state.
+
+    The tabular model is deterministic in drawing observations from the state,
+    in that given a certain state, the observation is always the same;
+    a vector with self.obs_dim entries.
+    """
+
+    def obs_from_state(self, state: int) -> np.ndarray:
```

Comment:
Yeah the fact that this is deterministic took me a while to understand because it did not make sense to me where observations were being drawn/generated. We should probably document this actually.

Reply:
Currently, the docstring in the base class says:
```
observation_matrix: 2-D array with observation probabilities for a
    given state, of shape `(n_states,n_observations)`.
```
This is why I was confused. `observation_matrix` is not actually a PMF, it's a mapping from a state to an observation vector. `n_observations` describes the dimensionality of the observation vector. Values can be negative and rarely sum to one. (see usage examples).

Reply:
RandomTransitionEnv is also relevant to this conversation. (there's a `random_obs: Whether to use random observations (True) or one-hot coded (False).`

Reply:
Yes, the observation matrix is random!

`RandomTransitionEnv` appears to use a *randomly generated* observation matrix but once generated it's fixed. The docstring there should perhaps be clarifying.

It's not necessarily a many-to-one mapping? It's just a deterministic mapping from state to observation.